### PR TITLE
enforce specification of vcs data for deployment whitelisting purposes

### PR DESCRIPTION
**Pull request by paulzah**
# Changes

* enforce ci_data as part of the expected payload if the `unique service names` feature is turned on

# Testing

* unit testing
* locally for regression 
* local payload validation test

### CODE REVIEW

File: slingshotapi/deploy.py [hunk 1]
```diff
@@ -64,6 +65,10 @@ def deploy(cls, deployment_data):
         engines = flask.current_app.config['SLINGSHOT_ENGINE'].get('endpoints', {})
         vcs_mappings = flask.current_app.config.get('VCS_MAPPINGS')
         repo_data = deployment_data.get("ci_data", {}).get("repo", {})
+        if flask.current_app.config.get('UNIQUE_SERVICE_NAMES') and repo_data is None:
```
Comment (gjtempleton) [inline slingshotapi/deploy.py]: `repo_data` will never be a `None` due to the line directly above surely?

**Comment by gnhenderson** [comment]
I don't think this is the correct way to be checking for whitelisting. What we want to do is make sure that the deployment request has come from a recognised vcs source.
The whitelist validation check needs to extract the source vcs url from the `['ci_data']['repo'][link_url']` and compare it to all the *values* in the `VCS_MAPPINGS` looking for a match.
If the vcs url in the payload is in the `VCS_MAPPINGS` dict the deploy can proceed otherwise the deploy request is rejected as the source VCS is not a recognised, i.e. whitelisted source.

File: slingshotapi/deploy.py (hunk 1 above)
Comment (paulzah) [inline slingshotapi/deploy.py]: :man_facepalming:

File: slingshotapi/deploy.py [hunk 2]
```diff
@@ -83,6 +84,10 @@ def deploy(cls, deployment_data):
                         if m else
                         "The VCS link url cannot be recognised: {}".format(link_url)
                     )
+            else:
+                raise cls.ParameterNotProvidedError(
```
Comment (balintzs) [inline slingshotapi/deploy.py]: Make sure, this error is handled in blueprints: https://github.com/Skyscanner/slingshot-api/blob/enforce_vcs_details/slingshotapi/blueprints/deployments.py#L14

Comment (paulzah) [inline slingshotapi/deploy.py]: good catch : pre-existing error class, assumed blueprint up-to-date - updating code now

**Comment by balintzs** [comment]
LGTM

**Pull request merged**